cardinal/ecs/gamestate/redis.go

```diff
@@ -13,10 +13,121 @@ import (
 	"pkg.world.dev/world-engine/cardinal/types/component"
 )
 
+type RedisStorage struct {
+	OtherClient   redis.Cmdable
+	CurrentClient redis.Cmdable
+}
+
+func (r *RedisStorage) GetFloat64(ctx context.Context, key string) (float64, error) {
+	res, err := r.CurrentClient.Get(ctx, key).Float64()
+	if err != nil {
+		return 0, eris.Wrap(err, "")
+	}
+	return res, nil
+}
+func (r *RedisStorage) GetFloat32(ctx context.Context, key string) (float32, error) {
+	res, err := r.CurrentClient.Get(ctx, key).Float32()
+	if err != nil {
+		return 0, eris.Wrap(err, "")
+	}
+	return res, nil
+}
+func (r *RedisStorage) GetUInt64(ctx context.Context, key string) (uint64, error) {
+	res, err := r.CurrentClient.Get(ctx, key).Uint64()
+	if err != nil {
+		return 0, eris.Wrap(err, "")
+	}
+	return res, nil
+}
+
+func (r *RedisStorage) GetInt64(ctx context.Context, key string) (int64, error) {
+	res, err := r.CurrentClient.Get(ctx, key).Int64()
+	if err != nil {
+		return 0, eris.Wrap(err, "")
+	}
+	return res, nil
+}
+
+func (r *RedisStorage) GetInt(ctx context.Context, key string) (int, error) {
+	res, err := r.CurrentClient.Get(ctx, key).Int()
+	if err != nil {
+		return 0, eris.Wrap(err, "")
+	}
+	return res, nil
+}
+
+func (r *RedisStorage) GetBool(ctx context.Context, key string) (bool, error) {
+	res, err := r.CurrentClient.Get(ctx, key).Bool()
+	if err != nil {
+		return false, eris.Wrap(err, "")
+	}
+	return res, nil
+}
+
+func (r *RedisStorage) GetBytes(ctx context.Context, key string) ([]byte, error) {
+	bz, err := r.CurrentClient.Get(ctx, key).Bytes()
+	if err != nil {
+		return nil, eris.Wrap(err, "")
+	}
+	return bz, nil
+}
+
+func (r *RedisStorage) Set(ctx context.Context, key string, value any) error {
+	return eris.Wrap(r.CurrentClient.Set(ctx, key, value, 0).Err(), "")
+}
+
+func (r *RedisStorage) Incr(ctx context.Context, key string) error {
+	return eris.Wrap(r.CurrentClient.Incr(ctx, key).Err(), "")
+}
+
+func (r *RedisStorage) Decr(ctx context.Context, key string) error {
+	return eris.Wrap(r.CurrentClient.Decr(ctx, key).Err(), "")
+}
+
+func (r *RedisStorage) Delete(ctx context.Context, key string) error {
+	return eris.Wrap(r.CurrentClient.Del(ctx, key).Err(), "")
+}
+
+func (r *RedisStorage) Close(ctx context.Context) error {
+	return eris.Wrap(r.CurrentClient.Shutdown(ctx).Err(), "")
+}
+
+func (r *RedisStorage) StartTransaction(_ context.Context) (Storage, error) {
+	pipeline := r.CurrentClient.TxPipeline()
+	redisTransaction := NewRedisStorage(pipeline)
+	redisTransaction.OtherClient = r.CurrentClient
+	return &redisTransaction, nil
+}
```

Comment:
couldn't we just use the same `RedisStorage` to start a transaction, rather than creating & returning a new one? seems like we could just flip clients internally

Reply:
Mutation is generally bad for coding. Should be avoided whenever possible as a rule of thumb. Better to create a new thing and the old thing retains the same behavior. This is my philosophy and it is how go-redis creates a transaction as well.

Specifically though for this use case there is an issue with out intuition on how storage is used: 

Imagine two references to the same RedisStorage, A and B. 
 ```
 A = &redisStorage.New()
 B = A
 
 process 1:
    B.write(3)
    
process 2: 
   A.startTransaction() // returns void transfroms B into transaction. 
   A.write(1)
   A.write(2)
   A.EndTransaction()
```

I expect 1 to be written after 2. So only these two outputs are possible:
1,2,3
3,1,2


But if A is changed to a transaction that could also change B and B can be part of the transaction leading to a potential race condition. We can now see 3 possible outputs: 
1,3,2
1,2,3
3,1,2

Which is against our intuition of how the transaction in process 2 should work. The transaction is supposed to block other writes. This mistake is prevented if startTransaction returns a new thing that represents a transaction. I'll re-emphasize that go-redis does this exact same pattern. I think the confusion here lies in the fact that I don't have a "Transaction" type I just reuse the RedisStorage type for the transaction. 

Type Aliasing could be used here if you want for clarity. I'm neutral about that. 
`type Transaction = Storage`

Reply:
As a side note. Rust semantics makes this bug I'm talking about here impossible from happening.

Reply:
@technicallyty can you take a look when you get the chance also prob better to talk in person today if you want.

Reply:
Sorry one other thing to add here. If we do it mutation style, there will have to be mutexes to make sure that mutating RedisStorage is thread safe. Doing it this way we just copy the reference address which is thread safe with no locking required.

Reply:
when i see `StartTransaction` i just don't really expect to get another copy of the storage object back. i think this would be nicer if we had a transactor interface or something. the whole client swapping seems a bit messy to me

Reply:
@technicallyty You are NOT getting a copy of Storage back. 

You are getting a transaction back. This is Exactly what go-redis does.  

It's just the Type of storage and type of transaction are identical. They have the same methods and the same interface this allows for things to operate on BOTH transaction and client in the future. We can use a type alias here if you want to separate concepts b

I can do the type alias here of:

`type Transaction = Storage`

this is actually deliberately crippling functions that accept storage as a parameter as those function can operate on both a transaction and a storage. It's solved by a quick typecast of `Transaction(storage)` or something of that nature so it's not a big deal to me. 

Would you prefer this?

Reply:
right, i know its a transaction you get back, but it LOOKS like you're getting just another storage object. my two main gripes here:

1. client swapping looks messy and confusing to me
2. `StartTransaction` should be `BranchedClient`, `TxClient` or something that indicates you're getting a new object back. the word `StartXXX` doesn't really seem like a method name that would return another object - which is why i initially suggested struct mutation (which i see now isn't the best route).

Reply:
also type aliasing won't necessarily cripple anything, because any functions that accept the `Storage` interface will still accept `Transaction` interface

Reply:
> also type aliasing won't necessarily cripple anything, because any functions that accept the `Storage` interface will still accept `Transaction` interface

You're right, I assumed there would have to be a cast here. But I just tried it, there's no way to type alias a generic type. Type aliasing is not a solution.

Reply:
No this works. @technicallyty I can do the above and I believe everything that accepts a transaction will accept a storage. Are you good with this?

Reply:
Also sorry, a later PR makes Storage generic. I was commenting on that.

Reply:
yeah i think that might be better, if @jerargus could sign off on that id say go for it. only concern i have left after that is the client juggling stuff. but im not really sure how to make it cleaner

Reply:
@jerargus I talked to tyler about it in my one on one. He told me he's going to talk to you about it I think he's relatively neutral about your suggestion. I personally disagree with it. 

I think it's easier to discuss this in person, I'll wait for tuesday or monday.

Reply:
talked with both @jerargus & @pyrofolium and it seems like the best path forward is:

- remove the double client business
- EndTransaction just returns an error. if someone needs the original storage, and the pipeline, they need to be explicit about it, and not expect a transaction to give them their OG client. (hopefully that makes sense).

I think aliasing the the storage interface for `Transaction` will still be a nice addition. just so everything is very clear-cut.

Reply:
> My view: The receiver for the StartTransaction method and the thing returned are fundamentally different things and should not be treated the same. Yes, the redis package says they are both Cmdable, but under the hood it doesn’t work like that. e.g. doing a read operation from a TxPipeline will fail.
> 
> I think the write methods from PrimitiveStorage should be moved into a separate interface (PrimitiveTransaction) which has Set, Incr, Decr, Delete, and EndTransaction. PrimitiveStorage has all the Get* methods, as well as StartTransaction and Close. StartTransaction would return a PrimitiveTransaction.
> 
> This also means the implementations of the methods should be split across 2 structs. One struct only knows about a redis.Client, and the other only knows about a redis.Pipeliner.

I'm combing over old reviews to clarify situations because in a unscheduled meeting with Tyler he mentioned my PRs have bad design and it's not "golike" enough so I want to find out where and why and pinpoint concretely the issues here.

Just want to comment here the above design written by jermey is actually I believe a mistaken design. PrimitiveStorage cannot have Set, Incr and Decr removed because the PrimitiveStorage interface needs to write data outside of a transaction. It's not supposed to be ONLY reads outside of a transaction and ONLY writes inside of a transaction. Correct me if I'm mistaken here but this was not your intention @jerargus? It's sort of a harmless error here and you likely meant something else. Just documenting this here, feel free to ignore.